src/persist-client/src/examples/persistent_source_example.rs

```diff
@@ -0,0 +1,1395 @@
+        /// _differential upper_, which is exclusive and denotes the frontier at which new updates
+        /// would happen. For reclocking puposes (finding/minting a matching timestamp binding) we
+        /// are interested in the largest from timestamps (think Kafka partition/offset pairs)
+        /// for which we have a binding.
+        inclusive_upper: Antichain<FT>,
+    }
+
+    impl<FT: timely::PartialOrder + Debug, TT: Copy + Debug> TimestampBindings<FT, TT> {
+        pub fn new(
+            lower: Antichain<FT>,
+            inclusive_upper: Antichain<FT>,
+            mut bindings: Vec<(FT, TT)>,
+        ) -> Self {
+            bindings.sort_by(|a, b| {
+                if a.0.less_than(&b.0) {
+                    std::cmp::Ordering::Less
+                } else if a.0.less_equal(&b.0) {
+                    std::cmp::Ordering::Equal
+                } else {
+                    std::cmp::Ordering::Greater
+                }
+            });
+            Self {
+                lower,
+                inclusive_upper,
+                bindings,
+            }
+        }
+
+        pub fn get_timestamp(&self, ts: FT) -> TT {
+            assert!(
+                self.lower.less_equal(&ts),
+                "from timestamp {:?} is not beyond lowers: {:?}",
+                ts,
+                self.lower
+            );
+            assert!(
+                !self.inclusive_upper.less_than(&ts),
+                "inclusive upper must not be less than ts"
+            );
+
+            // Very naive solution for this!
+
+            let mut result_ts = None;
+
+            for (from_ts, to_ts) in self.bindings.iter() {
+                if ts.less_equal(from_ts) {
+                    result_ts = Some(to_ts);
+                } else {
+                    break;
+                }
+            }
+
+            match result_ts {
+                Some(binding) => *binding,
+                None => panic!("missing binding for {:?}, bindings: {:?}", ts, self),
+            }
+        }
+    }
+
+    // A trait for doing iterated consensus.
+    //
+    // The updates are not differential, while they probably should be. As it is now, we can never
+    // retract updates.
+    #[async_trait]
+    pub trait IteratedConsensus<K, V, T> {
+        /// Returns all updates with timestamps up to (and including) the given `as_of`.
+        ///
+        /// Returns an `Err` with the current upper if the given `as_of` is beyond the current
+        /// upper.
+        async fn snapshot(
+            &self,
+            as_of: Antichain<T>,
+        ) -> Result<Result<Vec<((K, V), T)>, Antichain<T>>, ExternalError>;
+
+        /// Tries to append the given updates to the maintained collection. If the current
+        /// upper is not `expected_upper` this will return an `Err` containing the current
+        /// upper.
+        async fn compare_and_append(
```

Comment:
can this internally retry ExternalErrors so it doesn't have to return them?

Reply:
yep!